SAT/Z3Interfacing.hpp

```diff
@@ -141,19 +141,55 @@ class Z3Interfacing : public PrimitiveProofRecordingSATSolver
 
   struct FuncOrPredId
   {
-    explicit FuncOrPredId(unsigned id, bool isPredicate) : id(id), isPredicate(isPredicate) {}
-    explicit FuncOrPredId(Term* term) : FuncOrPredId(term->functor(), term->isLiteral()) {}
-    static FuncOrPredId function(FuncId id) { return FuncOrPredId ( id, false ); }
-    static FuncOrPredId predicate(PredId id) { return FuncOrPredId ( id, true ); }
+    explicit FuncOrPredId(unsigned id, bool isPredicate, Term *forSorts = nullptr) : id(id), isPredicate(isPredicate), forSorts(forSorts) {}
+    explicit FuncOrPredId(Term* term) :
+      FuncOrPredId(
+        term->functor(),
+        term->isLiteral(),
+        term->numTypeArguments() == 0
+          ? nullptr
+          : term
+      )
+    {}
+    static FuncOrPredId monomorphicFunction(FuncId id) { return FuncOrPredId (id, false); }
+    static FuncOrPredId monomorphicPredicate(PredId id) { return FuncOrPredId (id, true); }
     unsigned id;
     bool isPredicate;
+    /**
+     * polymorphic symbol application: treat e.g. f(<sorts>, ...) as f<sorts>(...) for Z3
+     * in the monomorphic case, nullptr
+     * in the polymorphic case, some term of the form f(<sorts>, ...) which we use only for sort information
+     */
+    Term *forSorts;
 
     friend struct std::hash<FuncOrPredId> ;
     friend bool operator==(FuncOrPredId const& l, FuncOrPredId const& r)
-    { return l.id == r.id && l.isPredicate == r.isPredicate; }
+    {
+      if(l.id != r.id || l.isPredicate != r.isPredicate)
+        return false;
+      if(!l.forSorts)
+        return true;
```

Comment:
If `forSorts` is always set, it should be possible to remove this check since in the case where `numTypeArguments() == 0` the for loop below will never be entered.